runtime/cudaq/distributed/builtin/mpi_comm_impl.cpp

```diff
@@ -0,0 +1,368 @@
+static int mpi_Bcast(const cudaqDistributedCommunicator_t *comm, void *buffer,
+                     int32_t count, DataType dataType, int32_t rootRank) {
+  return MPI_Bcast(buffer, count, convertType(dataType), rootRank,
+                   unpackMpiCommunicator(comm));
+}
+
+/// @brief Wrapper of MPI_Allreduce
+static int mpi_Allreduce(const cudaqDistributedCommunicator_t *comm,
+                         const void *sendBuffer, void *recvBuffer,
+                         int32_t count, DataType dataType, ReduceOp opType) {
+  if (opType == MIN_LOC) {
+    return MPI_Allreduce(sendBuffer, recvBuffer, count,
+                         convertTypeMinLoc(dataType), convertType(opType),
+                         unpackMpiCommunicator(comm));
+  } else {
+    return MPI_Allreduce(sendBuffer, recvBuffer, count, convertType(dataType),
+                         convertType(opType), unpackMpiCommunicator(comm));
+  }
+}
+
+/// @brief Wrapper of MPI_Allreduce with MPI_IN_PLACE
+static int mpi_AllreduceInplace(const cudaqDistributedCommunicator_t *comm,
+                                void *recvBuffer, int32_t count,
+                                DataType dataType, ReduceOp opType) {
+  return MPI_Allreduce(MPI_IN_PLACE, recvBuffer, count, convertType(dataType),
+                       convertType(opType), unpackMpiCommunicator(comm));
+}
+
+/// @brief Wrapper of MPI_Allgather
+static int mpi_Allgather(const cudaqDistributedCommunicator_t *comm,
+                         const void *sendBuffer, void *recvBuffer,
+                         int32_t count, DataType dataType) {
+  return MPI_Allgather(sendBuffer, count, convertType(dataType), recvBuffer,
+                       count, convertType(dataType),
+                       unpackMpiCommunicator(comm));
+}
+
+/// @brief Wrapper of MPI_Allgatherv
+static int mpi_AllgatherV(const cudaqDistributedCommunicator_t *comm,
+                          const void *sendBuf, int sendCount, void *recvBuf,
+                          const int *recvCounts, const int *displs,
+                          DataType dataType) {
+  return MPI_Allgatherv(sendBuf, sendCount, convertType(dataType), recvBuf,
+                        recvCounts, displs, convertType(dataType),
+                        unpackMpiCommunicator(comm));
+}
+
+/// @brief Wrapper of MPI_Isend and track pending requests for synchronization
+static int mpi_SendAsync(const cudaqDistributedCommunicator_t *comm,
+                         const void *buf, int count, DataType dataType,
+                         int peer, int32_t tag) {
+  if (PendingRequest::g_requests[comm].nActiveRequests == 2)
+    return -1;
+  MPI_Request *request =
+      &(PendingRequest::g_requests[comm]
+            .requests[PendingRequest::g_requests[comm].nActiveRequests]);
+  int res = MPI_Isend(buf, count, convertType(dataType), peer, tag,
+                      unpackMpiCommunicator(comm), request);
+  if (res != MPI_SUCCESS) {
+    MPI_Cancel(request);
```

Comment:
FWIW, I haven't had good success with `MPI_Cancel` actually successfully cancelling sends and receives correctly. MPI has too many options like "eager send", etc. that make that capability dubious. (I'm not asking for a change here ... I'm just making an unsolicited comment 😄 )